common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala

```diff
@@ -0,0 +1,316 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.common.metrics.sink
+
+import java.util
+import java.util.Properties
+import java.util.concurrent.ConcurrentLinkedQueue
+
+import scala.beans.BeanProperty
+
+import com.codahale.metrics.MetricRegistry
+import io.netty.channel.ChannelHandler.Sharable
+import org.apache.hadoop.shaded.com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility
+import org.apache.hadoop.shaded.com.fasterxml.jackson.annotation.PropertyAccessor
+import org.apache.hadoop.shaded.com.fasterxml.jackson.databind.ObjectMapper
+
+import org.apache.celeborn.common.metrics.{CelebornHistogram, CelebornTimer, ResettableSlidingWindowReservoir}
+import org.apache.celeborn.common.metrics.source.{AbstractSource, NamedCounter, NamedGauge, NamedHistogram, NamedTimer, Source}
+
+class MetricData(
+    @BeanProperty var name: String,
+    @BeanProperty var value: Any,
+    @BeanProperty var timestampMs: Long,
+    @BeanProperty var labelNames: util.List[String],
+    @BeanProperty var labelValues: util.List[String]) {}
+
+class JsonServlet(
+    val property: Properties,
+    val registry: MetricRegistry,
+    val sources: Seq[Source],
+    val servletPath: String) extends AbstractServlet(sources) {
+
+  val innerMetrics: ConcurrentLinkedQueue[MetricData] = new ConcurrentLinkedQueue[MetricData]
+
+  val mapper = new ObjectMapper()
+  mapper.setVisibility(PropertyAccessor.FIELD, Visibility.ANY)
+
+  override def getMetricsSnapshot: String = {
+    val metricDatas = new util.ArrayList[MetricData]
+    try {
+      sources.map(source => metricDatas.addAll(getMetrics(source)))
+      mapper.writeValueAsString(metricDatas)
+    } catch {
+      case e: Throwable =>
+        logError("failed to get json data for metrics", e)
+        mapper.writeValueAsString(new util.ArrayList[MetricData])
+    }
+  }
+
+  override def createHttpRequestHandler(): ServletHttpRequestHandler = {
+    new JsonHttpRequestHandler(servletPath, this)
+  }
+
+  override def stop(): Unit = {
+    innerMetrics.clear()
+  }
+
+  def getMetrics(source: Source): util.List[MetricData] = {
+    val absSource = source.asInstanceOf[AbstractSource]
+    absSource.counters().foreach(c => recordCounter(absSource, c))
+    absSource.gauges().foreach(g => recordGauge(absSource, g))
+    absSource.histograms().foreach(h => {
+      recordHistogram(absSource, h)
+      h.asInstanceOf[CelebornHistogram].reservoir
+        .asInstanceOf[ResettableSlidingWindowReservoir].reset()
+    })
+    absSource.timers().foreach(t => {
+      recordTimer(absSource, t)
+      t.timer.asInstanceOf[CelebornTimer].reservoir
+        .asInstanceOf[ResettableSlidingWindowReservoir].reset()
+    })
+    val metricDatas = new util.ArrayList[MetricData]
+    innerMetrics.synchronized {
+      while (!innerMetrics.isEmpty) {
+        metricDatas.add(innerMetrics.poll)
```

Comment:
When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

Reply:
> When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

I think it makes sense , and make some corrections.

Reply:
Thank you for your suggestion. I think it makes sense , and make some corrections.

> When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

@RexXiong  Thank you for your suggestion. I think it makes sense , and make some corrections.